### PR TITLE
Enable the Ethereum rate limiter

### PR DESCRIPTION
**Pull request by lukasz-zimnoch**
Refs: https://github.com/keep-network/keep-ecdsa/issues/524

Here we add the possibility of enabling the Ethereum rate limiter introduced by keep-network/keep-common#46

Rate-limiter wraps Ethereum contract backend. It can limit the average number of requests per second using the token bucket algorithm and the request concurrency using a semaphore.

### CODE REVIEW

File: pkg/chain/ethereum/connect.go [hunk 1]
```diff
@@ -83,6 +83,15 @@ func connectWithClient(
...
 ) (*ethereumChain, error) {
+	if config.RequestsPerSecondLimit > 0 || config.ConcurrencyLimit > 0 {
+		client = ethutil.WrapRateLimiting(
+			client,
+			&ethutil.RateLimiterConfig{
+				RequestsPerSecondLimit: config.RequestsPerSecondLimit,
+				ConcurrencyLimit:       config.ConcurrencyLimit,
+			},
+		).(*ethclient.Client)
+	}
```
[the remark this answers is not in the document]
Comment (lukasz-zimnoch) [reply pkg/chain/ethereum/connect.go]: My bad. Should work now.

File: pkg/chain/ethereum/connect.go [hunk 2]
```diff
@@ -175,6 +174,33 @@ func connectWithClient(
...
 
+func addClientWrappers(
+	config ethereum.Config,
+	backend bind.ContractBackend,
+) bind.ContractBackend {
+	loggingBackend := ethutil.WrapCallLogging(logger, backend)
+
+	if config.RequestsPerSecondLimit > 0 || config.ConcurrencyLimit > 0 {
+		logger.Infof(
+			"enabled ethereum rate limiter; "+
```
Comment (pdyraga) [inline pkg/chain/ethereum/connect.go:185]: `ethereum client request rate limiter` perhaps?

File: go.mod [hunk 3]
```diff
@@ -15,7 +15,7 @@ require (
 	github.com/ipfs/go-datastore v0.4.4
 	github.com/ipfs/go-log v1.0.4
 	github.com/keep-network/go-libp2p-bootstrap v0.0.0-20200423153828-ed815bc50aec
-	github.com/keep-network/keep-common v1.1.1-0.20200805115808-67d9cb633ddb
+	github.com/keep-network/keep-common v1.1.1-0.20200819140235-09eff2362ac0
```
Comment (pdyraga) [inline go.mod:18]: We have to update this one after merging https://github.com/keep-network/keep-core/pull/1959.

**Comment by pdyraga** [comment]
We need to update:
https://github.com/keep-network/keep-core/blob/2d3a9e0b70e80ac84a28addcb6ea24f2d5ce670b/solidity/scripts/lcl-client-config.js#L56

Otherwise, we'll end up with floating point values for `RequestsPerSecondLimit` and `ConcurrencyLimit`.

**Comment by pdyraga** [comment]
What are the recommended settings for `RequestsPerSecondLimit` and `ConcurrencyLimi` assuming I am running it against e.g. Infura? I am testing locally with 3 clients, each one holding about 22 group members, and configured with:
```
RequestsPerSecondLimit = 1
ConcurrencyLimit = 50
```

What I am observing is:
- relay request TX mined at block 1147 @ 14:39:10
- first "restoring signature from" log appeared around block 1190 @ 14:41:50
- last "restoring signature from"  log (on the same client 👆 ) appeared at 14:42:06
- relay request mined at block 1217 @ 14:43:41

**Comment by lukasz-zimnoch** [comment]
> What are the recommended settings for `RequestsPerSecondLimit` and `ConcurrencyLimi` assuming I am running it against e.g. Infura? I am testing locally with 3 clients, each one holding about 22 group members, and configured with:
> 
> ```
> RequestsPerSecondLimit = 1
> ConcurrencyLimit = 50
> ```
> 
> What I am observing is:
> 
>     * relay request TX mined at block 1147 @ 14:39:10
> 
>     * first "restoring signature from" log appeared around block 1190 @ 14:41:50
> 
>     * last "restoring signature from"  log (on the same client 👆 ) appeared at 14:42:06
> 
>     * relay request mined at block 1217 @ 14:43:41

Please take a look at https://github.com/keep-network/keep-core/pull/2014.

**Comment by pdyraga** [comment]
> What are the recommended settings for `RequestsPerSecondLimit` and `ConcurrencyLimi` assuming I am running it against e.g. Infura? I am testing locally with 3 clients, each one holding about 22 group members, and configured with:
> 
> ```
> RequestsPerSecondLimit = 1
> ConcurrencyLimit = 50
> ```
> 
> What I am observing is:
> 
>     * relay request TX mined at block 1147 @ 14:39:10
> 
>     * first "restoring signature from" log appeared around block 1190 @ 14:41:50
> 
>     * last "restoring signature from"  log (on the same client 👆 ) appeared at 14:42:06
> 
>     * relay request mined at block 1217 @ 14:43:41

We logged all calls the client is doing to Ethereum and it all makes sense and works as designed. Having 20 members on one client means that we are subscribing to relay entry submitted event 20 times and with 1 request per second limit, we have a delay of 20 seconds. Given that 1 request per second is not something anyone will use on prod and we shouldn't have 20 members on one client, it all works fine and we are not going to do anything about it now.

However, having the rate limiter lets us easily detect problematic spots - we found and already fixed one: https://github.com/keep-network/keep-core/pull/2014. We are going to look through all the calls client is doing separately and see if we can do some cuts, e.g subscribe one time and not for every member.

**Comment by pdyraga** [comment]
Tested with
```
RequestsPerSecondLimit = 10
ConcurrencyLimit = 10
```
as well as with
```
RequestsPerSecondLimit = 1
ConcurrencyLimit = 50
```

Works as advertised.

**Review by pdyraga — APPROVED** [review]

**Pull request merged**